File: 2020/cpp/day_06.cpp
#include "day_06.h"

#include <numeric>

int day_06::group_t::anyYes() const {
    std::array<bool, 26> isThere{};
    for (auto c : responses) {
        isThere[c - 'a'] = true;
    }
    return std::accumulate(std::begin(isThere), std::end(isThere), 0);
}

int day_06::group_t::allYes() const {
    std::array<int, 26> isThere{};
    for (auto c : responses) {
        ++isThere[c - 'a'];
    }
    int count{0};
    for (int i : isThere)
        count += i == number_of_responses;
    return count;
}

day_06::day_06() : input_strings{readFile("Inputs/day_06.txt")} {
    const auto start = std::chrono::high_resolution_clock::now();
    for (auto current_line = input_strings.begin(); current_line != input_strings.end();) {
        auto& group = input.emplace_back();
        if (*current_line == "")
            ++current_line;
        while (*current_line != "") {
            group.responses += *current_line;
            ++group.number_of_responses;
            ++current_line;
        }
    }
    const auto end = std::chrono::high_resolution_clock::now();
    fmt::print("Parsing input for day six took {}ns\n", std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count());
}

void day_06::part_one() const {
    const auto start = std::chrono::high_resolution_clock::now();

    long count{0};
    for (const auto& group : input) {
        count += group.anyYes();
    }

    const auto end = std::chrono::high_resolution_clock::now();
    fmt::print("The answer for day six part one is {}\n", count);
    fmt::print("Took {}ns\n", std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count());
}

void day_06::part_two() const {
    const auto start = std::chrono::high_resolution_clock::now();

    long count{0};
    for (const auto& group : input) {
        count += group.allYes();
    }

    const auto end = std::chrono::high_resolution_clock::now();
    fmt::print("The answer for day six part two is {}\n", count);
    fmt::print("Took {}ns\n", std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count());
}

File: 2020/rust/src/day06.rs
use std::time::Duration;
use std::time::Instant;

fn parse_input(input: &str) -> (Vec<Vec<&str>>, Duration) {
    let start = Instant::now();
    let collected = input
        .lines()
        .collect::<Vec<_>>()
        .split(|x| x.is_empty())
        .map(|s| s.to_vec())
        .collect();
    let end = Instant::now();
    (collected, end - start)
}

fn solve_part1(input: &[Vec<&str>]) -> (usize, Duration) {
    let start = Instant::now();
    let count = input
        .iter()
        .map(|group| {
            let mut is_there: [bool; 26] = Default::default();

            for person in group {
                for response in person.as_bytes() {
                    *is_there.get_mut((response - b'a') as usize).unwrap() = true;
                }
            }
            is_there.iter().filter(|x| **x).count()
        })
        .sum();
    let end = Instant::now();
    (count, end - start)
}

fn solve_part2(input: &[Vec<&str>]) -> (usize, Duration) {
    let start = Instant::now();
    let count = input
        .iter()
        .map(|group| {
            let mut responses: [usize; 26] = Default::default();
            for member in group {
                for response in member.as_bytes() {
                    responses[(response - b'a') as usize] += 1;
                }
            }
            responses.iter().filter(|i| **i == group.len()).count()
        })
        .sum();
    let end = Instant::now();
    (count, end - start)
}

pub fn run(input: String) {
    let (parsed_input, time) = parse_input(&input);
    println!("Day 06, parsing input took {}ns", time.as_nanos());
    let (p1, time) = solve_part1(&parsed_input);
    println!("Day 06, part 1: {}. Took {}ns", p1, time.as_nanos());
    let (p2, time) = solve_part2(&parsed_input);
    println!("Day 06, part 2: {}. Took {}ns", p2, time.as_nanos());
}

#[cfg(test)]
mod tests {
    use super::*;
    const INPUT: &str = "abc

a
b
c

ab
ac

a
a
a
a

b";

    #[test]
    fn day06_part1() {
        let parsed = parse_input(INPUT).0;
        assert_eq!(solve_part1(&parsed).0, 11);
    }
    #[test]
    fn day06_part2() {
        let parsed = parse_input(INPUT).0;
        assert_eq!(solve_part2(&parsed).0, 6);
    }
}
